### PR TITLE
Replace StateHolder.values with custom map

### PR DESCRIPTION
**Pull request by embeddedt**
Replace last usages of `StateHolder.values` map so we can null it out, saving 150MB or more in larger packs.

This needs a bit of testing with mods in case there are edge cases.

### CODE REVIEW

**Review by jpenilla — CHANGES_REQUESTED** [review]
Looks good but needs some minor style changes
- final on params and locals
- explicit this, even for inner classes
- avoid use of var

**Review by jpenilla — APPROVED** [review]

File: src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java [hunk 1]
```diff
@@ -7,12 +7,15 @@
...
 import org.spongepowered.asm.mixin.Final;
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Mutable;
 import org.spongepowered.asm.mixin.Overwrite;
 import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.Unique;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
```
Comment (Spottedleaf) [inline src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java:17]: remove this newline

File: src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java [hunk 2]
```diff
@@ -141,4 +152,24 @@ public <T extends Comparable<T>> T getValue(final Property<T> property) {
...
+    /**
+     * @reason Replace with optimisedTable
+     * @author embeddedt
+     */
+    @Overwrite
+    public Map<Property<?>, Comparable<?>> getValues() {
+        ZeroCollidingReferenceStateTable<O, S> table = this.optimisedTable;
+        // We have to use this.values until the table is loaded
+        return table.isLoaded() ? table.getMapView(this.tableIndex) : this.values;
```
Comment (Spottedleaf) [inline src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java:172]: table may be used as a map for property -> comparable even when not loaded, as the property index is initialised

File: src/main/java/ca/spottedleaf/moonrise/patches/blockstate_propertyaccess/util/ZeroCollidingReferenceStateTable.java [hunk 3]
```diff
@@ -151,7 +166,70 @@ public <T extends Comparable<T>> S trySet(final long index, final Property<T> pr
...
+        private final long stateIndex;
+        private EntrySet entrySet;
+
+        MapView(final long stateIndex) {
+            this.stateIndex = stateIndex;
+        }
+
+        @Override
+        public boolean containsKey(final Object key) {
+            return ZeroCollidingReferenceStateTable.this.properties.contains(key);
```
Comment (Spottedleaf) [inline src/main/java/ca/spottedleaf/moonrise/patches/blockstate_propertyaccess/util/ZeroCollidingReferenceStateTable.java:191]: use ZCRST#hasProperty

File: src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java (hunk 2 above)
Comment (embeddedt) [reply src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java:172]: In my testing, the table relies on `getValues` to initialize itself. If `getValues` unconditionally delegates to the table it ends up just not having any properties.

File: src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java [hunk 4]
```diff
@@ -141,4 +151,24 @@ public <T extends Comparable<T>> T getValue(final Property<T> property) {
...
         return property != null && this.optimisedTable.hasProperty(property);
     }
+
+    /**
+     * @reason Replace with optimisedTable
+     * @author embeddedt
+     */
+    @Overwrite
+    public Collection<Property<?>> getProperties() {
+        return this.optimisedTable.getProperties();
```
Comment (Spottedleaf) [inline src/main/java/ca/spottedleaf/moonrise/mixin/blockstate_propertyaccess/StateHolderMixin.java:161]: should mirror Vanilla by using unmodifiableCollection

File: src/main/java/ca/spottedleaf/moonrise/patches/blockstate_propertyaccess/util/ZeroCollidingReferenceStateTable.java [hunk 5]
```diff
@@ -4,20 +4,35 @@
 import ca.spottedleaf.moonrise.patches.blockstate_propertyaccess.PropertyAccess;
 import ca.spottedleaf.moonrise.patches.blockstate_propertyaccess.PropertyAccessStateHolder;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.AbstractObjectSet;
```
Comment (Spottedleaf) [inline src/main/java/ca/spottedleaf/moonrise/patches/blockstate_propertyaccess/util/ZeroCollidingReferenceStateTable.java:7]: clean up the imports
